src/r4300/new_dynarec/arm_cpu_features.c

```diff
@@ -0,0 +1,146 @@
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.          *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "arm_cpu_features.h"
+#include "api/callbacks.h"
+
+arm_cpu_features_t arm_cpu_features;
+
+const char procfile[] = "/proc/cpuinfo";
+
+static unsigned char check_arm_cpu_feature(const char* feature)
+{
+    unsigned char status = 0;
+    FILE *pFile = fopen (procfile, "r");
+    
+    if (pFile != NULL)
+    {
+        char line[1024];
+        
+        while (fgets (line , sizeof(line) , pFile) != NULL)
+        {
+            if (strncmp (line, "Features\t: ", 11))
+                continue;
+
+            if (strstr (line + 11, feature) != NULL)
+                status = 1;
+            
+            break;
+        }
+        fclose (pFile);
+    }
+    return status;
+}
+
+static unsigned char get_arm_cpu_implementer(void)
+{
+    int implementer = 0;
+    FILE *pFile = fopen (procfile, "r");
+    
+    if (pFile != NULL)
+    {
+        char line[1024];
+        
+        while (fgets (line , sizeof(line) , pFile) != NULL)
+        {
+            if (strncmp (line, "CPU implementer\t: ", 18))
+                continue;
+            
+            sscanf(line+18, "0x%02hhx", &implementer);
+            
+            break;
+        }
+        fclose (pFile);
+    }
+    return (unsigned char)implementer;
+}
+
+static unsigned short get_arm_cpu_part(void)
+{
+    int part = 0;
+    FILE *pFile = fopen (procfile, "r");
+    
+    if (pFile != NULL)
+    {
+        char line[1024];
+        
+        while (fgets (line , sizeof(line) , pFile) != NULL)
+        {
+            if (strncmp (line, "CPU part\t: ", 11))
+                continue;
+            
+            sscanf(line+11, "0x%03hx", &part);
+            
+            break;
+        }
+        fclose (pFile);
+    }
+    return (unsigned short)part;
+}
+
+void detect_arm_cpu_features(void)
+{
+    arm_cpu_features.SWP      = check_arm_cpu_feature("swp");
```

Comment:
So the file is open for each arm cpu feature? Meaning opening the file each time?

What about passing an array of char*? I just ask.